File: packages/typescript/test/errors.test.ts
import { describe, it, expect } from "vitest";
import {
  LeaksNowError,
  LeaksNowAuthError,
  LeaksNowQuotaError,
  LeaksNowValidationError,
  LeaksNowServerError,
  errorFromResponse,
} from "../src/errors.js";

describe("errorFromResponse", () => {
  it("maps 401/403 to LeaksNowAuthError", () => {
    expect(errorFromResponse(401, "Unauthorized", { m: 1 })).toBeInstanceOf(LeaksNowAuthError);
    expect(errorFromResponse(403, "Forbidden", null)).toBeInstanceOf(LeaksNowAuthError);
  });
  it("maps 429 to LeaksNowQuotaError and keeps retryAfterMs", () => {
    const err = errorFromResponse(429, "Too Many Requests", null, 2000);
    expect(err).toBeInstanceOf(LeaksNowQuotaError);
    expect(err.retryAfterMs).toBe(2000);
  });
  it("maps 400/422 to LeaksNowValidationError", () => {
    expect(errorFromResponse(400, "Bad Request", null)).toBeInstanceOf(LeaksNowValidationError);
    expect(errorFromResponse(422, "Unprocessable", null)).toBeInstanceOf(LeaksNowValidationError);
  });
  it("maps 5xx to LeaksNowServerError", () => {
    expect(errorFromResponse(500, "Server Error", null)).toBeInstanceOf(LeaksNowServerError);
    expect(errorFromResponse(503, "Unavailable", null)).toBeInstanceOf(LeaksNowServerError);
  });
  it("falls back to base LeaksNowError for other 4xx", () => {
    const err = errorFromResponse(418, "Teapot", null);
    expect(err).toBeInstanceOf(LeaksNowError);
    expect(err.constructor).toBe(LeaksNowError);
    expect(err.status).toBe(418);
  });
  it("base error carries code and body", () => {
    const err = new LeaksNowError("boom", { code: "timeout" });
    expect(err.code).toBe("timeout");
    expect(err.name).toBe("LeaksNowError");
  });
});

File: go/errors_test.go
package leaksnow

import (
	"errors"
	"testing"
	"time"
)

func TestErrorFromResponse(t *testing.T) {
	if err := errorFromResponse(401, "Unauthorized", nil, 0); !errors.As(err, new(*AuthError)) {
		t.Fatalf("401 should be *AuthError, got %T", err)
	}
	if err := errorFromResponse(403, "Forbidden", nil, 0); !errors.As(err, new(*AuthError)) {
		t.Fatalf("403 should be *AuthError, got %T", err)
	}
	var qe *QuotaError
	err := errorFromResponse(429, "Too Many Requests", nil, 2*time.Second)
	if !errors.As(err, &qe) {
		t.Fatalf("429 should be *QuotaError, got %T", err)
	}
	if qe.RetryAfter != 2*time.Second {
		t.Fatalf("RetryAfter = %v, want 2s", qe.RetryAfter)
	}
	if err := errorFromResponse(422, "Unprocessable", nil, 0); !errors.As(err, new(*ValidationError)) {
		t.Fatalf("422 should be *ValidationError, got %T", err)
	}
	if err := errorFromResponse(503, "Unavailable", nil, 0); !errors.As(err, new(*ServerError)) {
		t.Fatalf("503 should be *ServerError, got %T", err)
	}
	err = errorFromResponse(418, "Teapot", nil, 0)
	var base *Error
	if !errors.As(err, &base) || base.StatusCode != 418 {
		t.Fatalf("418 should be base *Error with status 418, got %T", err)
	}
	if errors.As(err, new(*AuthError)) {
		t.Fatalf("418 must not match *AuthError")
	}
}

func TestErrorMessage(t *testing.T) {
	e := &Error{StatusCode: 500, Status: "Server Error"}
	if e.Error() == "" {
		t.Fatal("Error() must not be empty")
	}
}
